server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -2086,19 +2083,20 @@ public static ClusterState updateWithSnapshots(ClusterState state,
                     final ShardId shardId = indexRoutingTable.shard(i).shardId();
                     final String shardRepoGeneration;
                     if (useShardGenerations) {
-                        if (isNewIndex) {
+                        final String inFlightGeneration = inFlightShardStates.bestGeneration(index, shardId.id(), shardGenerations);
+                        if (inFlightGeneration == null && isNewIndex) {
                             assert shardGenerations.getShardGen(index, shardId.getId()) == null
                                 : "Found shard generation for new index [" + index + "]";
                             shardRepoGeneration = ShardGenerations.NEW_SHARD_GEN;
                         } else {
-                            shardRepoGeneration = shardGenerations.getShardGen(index, shardId.getId());
+                            shardRepoGeneration = inFlightGeneration;
```

Comment:
Double checking my understanding, this will also update the shard generation to use for WAITING shard snapshots, right?

Reply:
Yes exactly, see a few line below where we use this variable:


```java
                    } else if (primary.relocating() || primary.initializing()) {
                        shardSnapshotStatus = new ShardSnapshotStatus(
                            primary.currentNodeId(), ShardState.WAITING, shardRepoGeneration);
```